batchconsumer/writer.go

```diff
@@ -0,0 +1,247 @@
+	b.rateLimiter = rate.NewLimiter(rate.Limit(b.config.ReadRateLimit), b.config.ReadBurstLimit)
+
+	b.startCheckpointListener(checkpointer, b.checkpointChan)
+
+	return nil
+}
+
+// handleCheckpointError returns true if checkout should be tried again.  Returns false otherwise.
+func (b *batchedWriter) handleCheckpointError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	cperr, ok := err.(kcl.CheckpointError)
+	if !ok {
+		b.log.ErrorD("unknown-checkpoint-error", kv.M{"msg": err.Error(), "shard-id": b.shardID})
+		return true
+	}
+
+	switch cperr.Error() {
+	case "ShutdownException": // Skips checkpointing
+		b.log.ErrorD("shutdown-checkpoint-exception", kv.M{
+			"msg": err.Error(), "shard-id": b.shardID,
+		})
+		return false
+	case "ThrottlingException":
+		b.log.ErrorD("checkpoint-throttle", kv.M{"shard-id": b.shardID})
+	case "InvalidStateException":
+		b.log.ErrorD("invalid-checkpoint-state", kv.M{"shard-id": b.shardID})
+	default:
+		b.log.ErrorD("checkpoint-error", kv.M{"shard-id": b.shardID, "msg": err})
+	}
+
+	return true
+}
+
+func (b *batchedWriter) startCheckpointListener(
+	checkpointer *kcl.Checkpointer, checkpointChan <-chan batcher.SequencePair,
+) {
+	lastCheckpoint := time.Now()
+
+	go func() {
+		for {
+			seq := <-checkpointChan
+
+			// This is a write throttle to ensure we don't checkpoint faster than
+			// b.config.CheckpointFreq.  The latest seq number is always used.
+			for time.Now().Sub(lastCheckpoint) < b.config.CheckpointFreq {
+				select {
+				case seq = <-checkpointChan: // Keep updating checkpoint seq while waiting
+				case <-time.NewTimer(b.config.CheckpointFreq - time.Now().Sub(lastCheckpoint)).C:
+				}
+			}
+
+			retry := true
+			for n := 0; retry && n < b.config.CheckpointRetries+1; n++ {
+				str := seq.Sequence.String()
+				err := checkpointer.Checkpoint(&str, &seq.SubSequence)
+				if err == nil { // Successfully checkpointed!
+					lastCheckpoint = time.Now()
+					break
+				}
+
+				retry = b.handleCheckpointError(err)
+
+				if n == b.config.CheckpointRetries {
+					b.log.ErrorD("checkpoint-retries", kv.M{"attempts": b.config.CheckpointRetries})
+					retry = false
+				}
+
+				if retry {
+					time.Sleep(b.config.CheckpointRetrySleep)
+				}
+			}
+		}
+	}()
+}
+
+func (b *batchedWriter) createBatcher(tag string) batcher.Batcher {
+	sync := &batcherSync{
+		tag:    tag,
+		writer: b,
+	}
+	return batcher.New(sync, b.config.FlushInterval, b.config.FlushCount, b.config.FlushSize)
+}
+
+func (b *batchedWriter) splitMessageIfNecessary(record []byte) ([][]byte, error) {
+	// We handle two types of records:
+	// - records emitted from CWLogs Subscription
+	// - records emiited from KPL
+	if !splitter.IsGzipped(record) {
+		// Process a single message, from KPL
+		return [][]byte{record}, nil
+	}
+
+	// Process a batch of messages from a CWLogs Subscription
+	return splitter.GetMessagesFromGzippedInput(record, b.config.DeployEnv == "production")
+}
```

Comment:
Originally the `splitter` was to split logs from AWS Batch, but it sounds like we might be removing the splitter logic to a separate app.  Is that true @nathanleiby?